File: src/wrappers/python/samples/sign_bbs04.py
#!/usr/bin/env python3
from pygroupsig import groupsig
from pygroupsig import signature
from pygroupsig import memkey
from pygroupsig import grpkey
from pygroupsig import mgrkey
from pygroupsig import gml
from pygroupsig import identity
from pygroupsig import constants

# Setup
bbs04 = groupsig.setup(constants.BBS04_CODE)
gpk = bbs04['grpkey']
msk = bbs04['mgrkey']
gml = bbs04['gml']

# Join
msg1 = groupsig.join_mgr(0, msk, gpk, gml = gml)
msg2 = groupsig.join_mem(1, gpk, msgin = msg1)
usk = msg2['memkey']

# Sign
sig = groupsig.sign("Hello, World!", usk, gpk)

# Verify
b = groupsig.verify(sig, "Hello, World!", gpk)

if b == True:
    print ("VALID signature.")
else:
    print ("WRONG signature.")
    sys.exit()

# Open
id = groupsig.open(sig, msk, gpk, gml)
str = identity.identity_to_string(id)

print("Identity: %s" % str)

groupsig.clear(constants.BBS04_CODE, bbs04['config'])

File: src/wrappers/nodejs/samples/sign_bbs04.js
#!/usr/bin/env nodejs

const jsgroupsig = require('jsgroupsig');

/* Init groupsig */
jsgroupsig.init(jsgroupsig.BBS04);

/* Init grp_key */
let grpkey = jsgroupsig.grp_key_init(jsgroupsig.BBS04);

/* Init manager key */
let mgrkey = jsgroupsig.mgr_key_init(jsgroupsig.BBS04);

/* Init GML */
let gml = jsgroupsig.gml_init(jsgroupsig.BBS04);

/* Setup call: initializes group key and manager key */
jsgroupsig.setup(jsgroupsig.BBS04, grpkey, mgrkey, gml);

/* Add a member */
let memkey = jsgroupsig.mem_key_init(jsgroupsig.BBS04);
let msg1 = jsgroupsig.join_mgr(0, mgrkey, grpkey, null, gml);
let msg2 = jsgroupsig.join_mem(1, memkey, grpkey, msg1);

/* sign */
let sig = jsgroupsig.sign("Hello, World!", memkey, grpkey);

/* verify */
let ok = jsgroupsig.verify(sig, "Hello, World!", grpkey);

if (ok) console.log("VALID signature.");
else console.log("WRONG signature.");

/* Open the signature */
let id = jsgroupsig.open(sig, grpkey, mgrkey, gml);
let { index, proof } = jsgroupsig.open(sig, grpkey, mgrkey, gml);
console.log("Signer was: "+index);

/* Free stuff */
jsgroupsig.gml_free(gml);
jsgroupsig.grp_key_free(grpkey);
jsgroupsig.mgr_key_free(mgrkey);
jsgroupsig.mem_key_free(memkey);
jsgroupsig.clear(jsgroupsig.BBS04);
